x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/accesscontrol/OptOutQueryCache.java

```diff
@@ -63,24 +96,47 @@ public Weight doCache(Weight weight, QueryCachingPolicy policy) {
      */
     static boolean cachingIsSafe(Weight weight, IndicesAccessControl.IndexAccessControl permissions) {
         // support caching for common queries, by inspecting the field
-        // TODO: If in the future there is a Query#extractFields() then we can do a better job
-        Set<String> fields = new HashSet<>();
         try {
-            FieldExtractor.extractFields(weight.getQuery(), fields);
-        } catch (UnsupportedOperationException ok) {
-            // we don't know how to safely extract the fields of this query, don't cache.
-            return false;
-        }
+            weight.getQuery().visit(new QueryVisitor() {
```

Comment:
Should we also check whether `weight.getQuery()` itself is in the allowlist?

Reply:
I think it's covered by the checks in `#consumeTerms` and `#visitLeaf`.

Reply:
I thought the assumption here is that we might not want to completely trust the `visit` implementation? Well, I think the assumption is theoretically a good one but does not really work in practice because we cannot completely avoid trusting it which goes back to Luca's [comment](https://github.com/elastic/elasticsearch/pull/93390#discussion_r1112136338).